lib/concurrent/atomic/thread_local_var.rb

```diff
@@ -33,169 +33,131 @@ module Concurrent
   #
   #   @see https://docs.oracle.com/javase/7/docs/api/java/lang/ThreadLocal.html Java ThreadLocal
   #
-  # @!visibility private
-  class AbstractThreadLocalVar
+  class ThreadLocalVar
+
+    # Each thread has a (lazily initialized) array of thread-local variable values
+    # Each time a new thread-local var is created, we allocate an "index" for it
+    # For example, if the allocated index is 1, that means slot #1 in EVERY
+    #   thread's thread-local array will be used for the value of that TLV
+    #
+    # The good thing about using a per-THREAD structure to hold values, rather
+    #   than a per-TLV structure, is that no synchronization is needed when
+    #    reading and writing those values (since the structure is only ever
+    #    accessed by a single thread)
+    #
+    # Of course, when a TLV is GC'd, 1) we need to recover its index for use
+    #   by other new TLVs (otherwise the thread-local arrays could get bigger
+    #   and bigger with time), and 2) we need to null out all the references
+    #   held in the now-unused slots (both to avoid blocking GC of those objects,
+    #   and also to prevent "stale" values from being passed on to a new TLV
+    #   when the index is reused)
+    # Because we need to null out freed slots, we need to keep references to
+    #   ALL the thread-local arrays -- ARRAYS is for that
+    # But when a Thread is GC'd, we need to drop the reference to its thread-local
+    #   array, so we don't leak memory
 
     # @!visibility private
     NIL_SENTINEL = Object.new
-    private_constant :NIL_SENTINEL
+    FREE = []
+    LOCK = Mutex.new
+    ARRAYS = {} # used as a hash set
+    @@next = 0
 
-    # @!macro [attach] thread_local_var_method_initialize
-    #
     #   Creates a thread local variable.
     #
     #   @param [Object] default the default value when otherwise unset
     def initialize(default = nil)
       @default = default
-      allocate_storage
+      @index = LOCK.synchronize do
+        FREE.pop || begin
+          result = @@next
+          @@next += 1
+          result
+        end
+      end
+      ObjectSpace.define_finalizer(self, self.class.threadlocal_finalizer(@index))
     end
 
-    # @!macro [attach] thread_local_var_method_get
-    #
     #   Returns the value in the current thread's copy of this thread-local variable.
     #
     #   @return [Object] the current value
     def value
-      value = get
-
-      if value.nil?
-        @default
-      elsif value == NIL_SENTINEL
-        nil
+      if array = Thread.current[:__threadlocal_array__]
+        value = array[@index]
+        if value.nil?
+          @default
+        elsif value.equal?(NIL_SENTINEL)
```

Comment:
So do you think NO_VALUE would be a more appropriate name? Or stick with
NIL_SENTINEL?

On Thu, Jul 23, 2015 at 4:52 PM, jdantonio <notifications@github.com> wrote:

> In lib/concurrent/atomic/thread_local_var.rb
> <https://github.com/ruby-concurrency/concurrent-ruby/pull/376#discussion_r35329659>
> :
>
> >      #   Returns the value in the current thread's copy of this thread-local variable.
> >      #
> >      #   @return [Object] the current value
> >      def value
> > -      value = get
> > -
> > -      if value.nil?
> > -        @default
> > -      elsif value == NIL_SENTINEL
> > -        nil
> > +      if array = Thread.current[:__threadlocal_array__]
> > +        value = array[@index]
> > +        if value.nil?
> > +          @default
> > +        elsif value.equal?(NIL_SENTINEL)
>
> Ah, I see it now. This is what I get for reviewing a PR while on a
> conference call. :-) I saw NIL_SENTINEL here and--in my mind--registered
> that it was the return value. We use NO_VALUE as a return value in other
> places in this gem. Sorry for my confusion.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ruby-concurrency/concurrent-ruby/pull/376/files#r35329659>
> .
>